pkg/instancetype/find/spec_test.go

```diff
@@ -0,0 +1,350 @@
+}
+
+var _ = Describe("Preference SpecFinder", func() {
+	var (
+		finder instancetypeSpecFinder
+		vm     *v1.VirtualMachine
+
+		virtClient                       *kubecli.MockKubevirtClient
+		fakeClientset                    *fake.Clientset
+		instancetypeInformerStore        cache.Store
+		clusterInstancetypeInformerStore cache.Store
+		controllerRevisionInformerStore  cache.Store
+	)
+
+	BeforeEach(func() {
+		ctrl := gomock.NewController(GinkgoT())
+		virtClient = kubecli.NewMockKubevirtClient(ctrl)
+
+		virtClient.EXPECT().AppsV1().Return(k8sfake.NewSimpleClientset().AppsV1()).AnyTimes()
+
+		fakeClientset = fake.NewSimpleClientset()
+
+		virtClient.EXPECT().VirtualMachine(metav1.NamespaceDefault).Return(
+			fakeClientset.KubevirtV1().VirtualMachines(metav1.NamespaceDefault)).AnyTimes()
+
+		virtClient.EXPECT().VirtualMachineInstancetype(metav1.NamespaceDefault).Return(
+			fakeClientset.InstancetypeV1beta1().VirtualMachineInstancetypes(metav1.NamespaceDefault)).AnyTimes()
+
+		virtClient.EXPECT().VirtualMachineClusterInstancetype().Return(
+			fakeClientset.InstancetypeV1beta1().VirtualMachineClusterInstancetypes()).AnyTimes()
+
+		instancetypeInformer, _ := testutils.NewFakeInformerFor(&v1beta1.VirtualMachineInstancetype{})
+		instancetypeInformerStore = instancetypeInformer.GetStore()
+
+		clusterInstancetypeInformer, _ := testutils.NewFakeInformerFor(&v1beta1.VirtualMachineClusterInstancetype{})
+		clusterInstancetypeInformerStore = clusterInstancetypeInformer.GetStore()
+
+		controllerRevisionInformer, _ := testutils.NewFakeInformerFor(&appsv1.ControllerRevision{})
+		controllerRevisionInformerStore = controllerRevisionInformer.GetStore()
+
+		finder = find.NewSpecFinder(
+			instancetypeInformerStore,
+			clusterInstancetypeInformerStore,
+			controllerRevisionInformerStore,
+			virtClient,
+		)
+	})
+
+	It("find returns nil when no instancetype is specified", func() {
+		vm = libvmi.NewVirtualMachine(libvmi.New())
+		spec, err := finder.Find(vm)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(spec).To(BeNil())
+	})
+
+	It("find returns error when invalid Instancetype Kind is specified", func() {
+		vm = libvmi.NewVirtualMachine(libvmi.New())
+		vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+			Name: "foo",
+			Kind: "bar",
+		}
+		spec, err := finder.Find(vm)
+		Expect(err).To(MatchError(ContainSubstring("got unexpected kind in InstancetypeMatcher")))
+		Expect(spec).To(BeNil())
+	})
+
+	Context("Using global ClusterInstancetype", func() {
+		var clusterInstancetype *v1beta1.VirtualMachineClusterInstancetype
+
+		BeforeEach(func() {
+			clusterInstancetype = &v1beta1.VirtualMachineClusterInstancetype{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "test-cluster-instancetype",
+				},
+				Spec: v1beta1.VirtualMachineInstancetypeSpec{
+					CPU: v1beta1.CPUInstancetype{
+						Guest: uint32(2),
+					},
+					Memory: v1beta1.MemoryInstancetype{
+						Guest: resource.MustParse("128Mi"),
+					},
+				},
+			}
```

Comment:
For later: It would be nice if we could move `libinstancetype` to `pkg/`. :)

Reply:
I was thinking about moving the builder to `pkg/instancetype/builder` FWIW but `pkg/libinstancetype` would be more in line with `pkg/libvmi` I guess.